File: src/com/aopphp/go/PointcutQuerySyntaxHighlighter.java
package com.aopphp.go;


import com.aopphp.go.parser.PointcutLexer;
import com.intellij.lexer.FlexAdapter;
import com.intellij.lexer.Lexer;
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors;
import com.intellij.openapi.editor.colors.TextAttributesKey;
import com.intellij.openapi.editor.markup.TextAttributes;
import com.intellij.openapi.fileTypes.SyntaxHighlighterBase;
import com.intellij.psi.TokenType;
import com.intellij.psi.tree.IElementType;
import com.aopphp.go.psi.PointcutTypes;
import com.intellij.ui.JBColor;
import org.jetbrains.annotations.NotNull;

import java.awt.*;
import java.io.Reader;

import static com.intellij.openapi.editor.colors.TextAttributesKey.createTextAttributesKey;

public class PointcutQuerySyntaxHighlighter extends SyntaxHighlighterBase {
    public static final TextAttributesKey SEPARATOR =
            createTextAttributesKey("GO_POINTCUT_SEPARATOR", DefaultLanguageHighlighterColors.OPERATION_SIGN);
    public static final TextAttributesKey KEYWORD =
            createTextAttributesKey("GO_POINTCUT_KEYWORD", DefaultLanguageHighlighterColors.KEYWORD);
    public static final TextAttributesKey COMMENT =
            createTextAttributesKey("GO_POINTCUT_COMMENT", DefaultLanguageHighlighterColors.LINE_COMMENT);
    public static final TextAttributesKey PARENTHESIS =
            createTextAttributesKey("GO_POINTCUT_PARENTHESIS", DefaultLanguageHighlighterColors.PARENTHESES);
    public static final TextAttributesKey MEMBER_MODIFIER =
            createTextAttributesKey("GO_POINTCUT_MEMBER_MODIFIER", DefaultLanguageHighlighterColors.STRING);

    static final TextAttributesKey BAD_CHARACTER = createTextAttributesKey("GO_POINTCUT_BAD_CHARACTER",
            new TextAttributes(JBColor.RED, null, null, null, Font.BOLD));

    private static final TextAttributesKey[] BAD_CHAR_KEYS = new TextAttributesKey[]{BAD_CHARACTER};
    private static final TextAttributesKey[] SEPARATOR_KEYS = new TextAttributesKey[]{SEPARATOR};
    private static final TextAttributesKey[] KEY_KEYS = new TextAttributesKey[]{KEYWORD};
    private static final TextAttributesKey[] COMMENT_KEYS = new TextAttributesKey[]{COMMENT};
    private static final TextAttributesKey[] PARENTHESIS_KEYS = new TextAttributesKey[]{PARENTHESIS};
    private static final TextAttributesKey[] MEMBER_MODIFIER_KEYS = new TextAttributesKey[]{MEMBER_MODIFIER};
    private static final TextAttributesKey[] EMPTY_KEYS = new TextAttributesKey[0];

    @NotNull
    @Override
    public Lexer getHighlightingLexer() {
        return new FlexAdapter(new PointcutLexer((Reader) null));
    }

    @NotNull
    @Override
    public TextAttributesKey[] getTokenHighlights(IElementType tokenType) {
        if (tokenType.equals(PointcutTypes.T_NS_SEPARATOR)
            || tokenType.equals(PointcutTypes.T_ASTERISK)
            || tokenType.equals(PointcutTypes.T_STATIC_ACCESS)
            || tokenType.equals(PointcutTypes.T_OBJECT_ACCESS)
            || tokenType.equals(PointcutTypes.T_SUBNAMESPACE_SIGN)) {
            return SEPARATOR_KEYS;
        } else if (tokenType.equals(PointcutTypes.ACCESS)
            || tokenType.equals(PointcutTypes.EXECUTION)
            || tokenType.equals(PointcutTypes.WITHIN)
            || tokenType.equals(PointcutTypes.INITIALIZATION)
            || tokenType.equals(PointcutTypes.STATICINITIALIZATION)
            || tokenType.equals(PointcutTypes.CFLOWBELOW)
            || tokenType.equals(PointcutTypes.DYNAMIC)) {
            return KEY_KEYS;
        } else if (tokenType.equals(PointcutTypes.T_LEFT_PAREN)
            || tokenType.equals(PointcutTypes.T_RIGHT_PAREN)) {
            return PARENTHESIS_KEYS;
        } else if (tokenType.equals(PointcutTypes.PRIVATE)
            || tokenType.equals(PointcutTypes.PROTECTED)
            || tokenType.equals(PointcutTypes.PUBLIC)
            || tokenType.equals(PointcutTypes.FINAL)) {
            return MEMBER_MODIFIER_KEYS;
        } else if (tokenType.equals(PointcutTypes.T_COMMENT)) {
            return COMMENT_KEYS;
        } else if (tokenType.equals(TokenType.BAD_CHARACTER)) {
            return BAD_CHAR_KEYS;
        } else {
            return EMPTY_KEYS;
        }
    }
}

File: src/main/kotlin/com/aopphp/go/PointcutQuerySyntaxHighlighter.kt
package com.aopphp.go

import com.aopphp.go.parser.PointcutLexer
import com.aopphp.go.psi.PointcutTypes
import com.intellij.lexer.FlexAdapter
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors
import com.intellij.openapi.editor.HighlighterColors
import com.intellij.openapi.editor.colors.TextAttributesKey
import com.intellij.openapi.editor.colors.TextAttributesKey.createTextAttributesKey
import com.intellij.openapi.fileTypes.SyntaxHighlighterBase
import com.intellij.psi.TokenType
import com.intellij.psi.tree.IElementType

class PointcutQuerySyntaxHighlighter : SyntaxHighlighterBase() {

    companion object {
        @JvmField val SEPARATOR       = createTextAttributesKey("GO_POINTCUT_SEPARATOR",       DefaultLanguageHighlighterColors.OPERATION_SIGN)
        @JvmField val KEYWORD         = createTextAttributesKey("GO_POINTCUT_KEYWORD",          DefaultLanguageHighlighterColors.KEYWORD)
        @JvmField val COMMENT         = createTextAttributesKey("GO_POINTCUT_COMMENT",          DefaultLanguageHighlighterColors.LINE_COMMENT)
        @JvmField val PARENTHESIS     = createTextAttributesKey("GO_POINTCUT_PARENTHESIS",      DefaultLanguageHighlighterColors.PARENTHESES)
        @JvmField val MEMBER_MODIFIER = createTextAttributesKey("GO_POINTCUT_MEMBER_MODIFIER",  DefaultLanguageHighlighterColors.KEYWORD)
        @JvmField val CLASS_REFERENCE = createTextAttributesKey("GO_POINTCUT_CLASS_REFERENCE",  DefaultLanguageHighlighterColors.CLASS_REFERENCE)
        @JvmField val BAD_CHARACTER   = createTextAttributesKey("GO_POINTCUT_BAD_CHARACTER",  HighlighterColors.BAD_CHARACTER)
    }

    override fun getHighlightingLexer() = FlexAdapter(PointcutLexer(null))

    override fun getTokenHighlights(tokenType: IElementType): Array<TextAttributesKey> = when (tokenType) {
        PointcutTypes.T_ANNOTATION,
        PointcutTypes.T_NS_SEPARATOR,
        PointcutTypes.T_ASTERISK,
        PointcutTypes.T_STATIC_ACCESS,
        PointcutTypes.T_OBJECT_ACCESS,
        PointcutTypes.T_SUBNAMESPACE_SIGN,
        PointcutTypes.T_COLON,
        PointcutTypes.T_QUESTION_MARK -> arrayOf(SEPARATOR)

        PointcutTypes.ACCESS,
        PointcutTypes.EXECUTION,
        PointcutTypes.WITHIN,
        PointcutTypes.INITIALIZATION,
        PointcutTypes.STATICINITIALIZATION,
        PointcutTypes.CFLOWBELOW,
        PointcutTypes.DYNAMIC,
        PointcutTypes.MATCHINHERITED -> arrayOf(KEYWORD)

        PointcutTypes.T_LEFT_PAREN,
        PointcutTypes.T_RIGHT_PAREN -> arrayOf(PARENTHESIS)

        PointcutTypes.PRIVATE,
        PointcutTypes.PROTECTED,
        PointcutTypes.PUBLIC,
        PointcutTypes.FINAL -> arrayOf(MEMBER_MODIFIER)

        PointcutTypes.T_COMMENT -> arrayOf(COMMENT)
        TokenType.BAD_CHARACTER -> arrayOf(BAD_CHARACTER)
        else -> emptyArray()
    }
}
